database/storage_engine.c

```diff
@@ -26,47 +32,62 @@ static STORAGE_ENGINE engines[] = {
         .id = RRD_MEMORY_MODE_NONE,
         .name = RRD_MEMORY_MODE_NONE_NAME,
         .api = {
+            .engine_ops = engine_ops_null,
             .collect_ops = im_collect_ops,
             .query_ops = im_query_ops
-        }
+        },
+        .context = NULL
     },
     {
         .id = RRD_MEMORY_MODE_RAM,
         .name = RRD_MEMORY_MODE_RAM_NAME,
         .api = {
+            .engine_ops = engine_ops_null,
             .collect_ops = im_collect_ops,
             .query_ops = im_query_ops
-        }
+        },
+        .context = NULL
     },
     {
         .id = RRD_MEMORY_MODE_MAP,
         .name = RRD_MEMORY_MODE_MAP_NAME,
         .api = {
+            .engine_ops = engine_ops_null,
             .collect_ops = im_collect_ops,
             .query_ops = im_query_ops
-        }
+        },
+        .context = NULL
     },
     {
         .id = RRD_MEMORY_MODE_SAVE,
         .name = RRD_MEMORY_MODE_SAVE_NAME,
         .api = {
+            .engine_ops = engine_ops_null,
             .collect_ops = im_collect_ops,
             .query_ops = im_query_ops
-        }
+        },
+        .context = NULL
     },
     {
         .id = RRD_MEMORY_MODE_ALLOC,
         .name = RRD_MEMORY_MODE_ALLOC_NAME,
         .api = {
+            .engine_ops = engine_ops_null,
             .collect_ops = im_collect_ops,
             .query_ops = im_query_ops
-        }
+        },
+        .context = NULL
     },
 #ifdef ENABLE_DBENGINE
     {
         .id = RRD_MEMORY_MODE_DBENGINE,
         .name = RRD_MEMORY_MODE_DBENGINE_NAME,
         .api = {
+            .engine_ops = {
+                .create = rrdeng_init,
+                .exit = rrdeng_prepare_exit,
+                .destroy = rrdeng_exit
```

Comment:
database/storage_engine.c:89:28: warning: initialization of 'void (*)(STORAGE_ENGINE_INSTANCE *)' {aka 'void (*)(struct storage_engine_instance *)'} from incompatible pointer type 'int (*)(STORAGE_ENGINE_INSTANCE *)' {aka 'int (*)(struct storage_engine_instance *)'} [-Wincompatible-pointer-types]
   89 |                 .destroy = rrdeng_exit
      |                            ^~~~~~~~~~~
database/storage_engine.c:89:28: note: (near initialization for 'engines[5].api.engine_ops.destroy')